### DISCUSSION TITLE
Philosophy & Direction

### DISCUSSION THREAD

**Discussion by ixxie** [litdb/litdb#5] [General]
Hey there,

I've just spotted your project and its very exciting! I'm seriously considering it for my project, despite being an SQL lover and great skeptic of ORMs. But since its quite early I want to get a sense of where your project is going, i.e. what is the eventual scope of the package (regardless of _when_ this will happen).

The more general question is regarding **migrations** - are you planning any specific support for migrations? If so what would be the approach? I've been curious about the possibility of having a clean approach to support _both_ a declarative schema definition _and_ a migration pathway and I was wondering if you had thoughts about this.

The more esoteric thought I've been having is about what I would called **type-safe de/serialization**. Typescript types are obviously far more expressive than any of the SQL dialects you support. Often we would serialize some type to store it in the database and than deserialize it  when retrieving it. I've recently started to wonder if a typescript SQL interface _should_ support this sort of thing out of the box: allow you to define _both_ the SQL type _and_ the Typescript type, by allowing you to define de/serialization functions to cast between them. Is this sane? Is this out of scope for litdb?

Finally, I wanted to point out that `$` is reserved in [Svelte](https://svelte.dev/). Of course, this is not as major a framework as React or Vue, but it is quite popular. I know its not _that_ big of a deal since one can easily substitute the symbol, but it means every example needs to be modified before use.

Thanks for your awesome work!

**Comment by mythz** [comment]
> But since its quite early I want to get a sense of where your project is going, i.e. what is the eventual scope of the package

Right it's still early early days, but to get a general idea of how I think an ORM/DB Abstraction library should be designed you can take a look at our mature .NET ORM https://docs.servicestack.net/ormlite/ which litdb is the spiritual port of.

The primary difference being C# is statically typed and its compiler gives us access to the LINQ Expressions AST at runtime which allows translating C# code expressions into SQL. We don't have the same LINQ-like AST in JS but TypeScript is a more powerful type system where we can achieve a similar SQL-Like "Typed Expression" using tagged template literals and TypeScript's amazing type system which allows us the freedom of a SQL Expression string to create any SQL we want, but also contains typed references to our App's data models allowing us to leverage TypeScript's type system to assert queries can only reference Tables included in the query and only reference props on our table data models whilst ensuring any values are parameterized.

The "philosophy" in both ORMs is to keep a close affinity to SQL where classes/properties have a 1:1 mapping to Tables/Columns and all APIs on query builders are named after the SQL construct they implement so that it's intuitive to know both how to create queries and know what SQL a query generates.

> The more general question is regarding migrations - are you planning any specific support for migrations? If so what would be the approach? 

That would depend on the usage of litdb and the interest in implementing that feature request. To see it implemented sooner feel free to submit a feature request for it at https://github.com/litdb/litdb/discussions/categories/ideas so I can measure interest in the feature and will let you get notified when it gets implemented.

If I do get around to implementing it, it will most likely adopt the same approach we took in our [OrmLite's Code First DB Migrations](https://docs.servicestack.net/ormlite/db-migrations) where you could use either [declarative](https://docs.servicestack.net/ormlite/db-migrations#declarative-code-first-migrations) and [imperative migrations](https://docs.servicestack.net/ormlite/db-migrations#imperative-migrations). The typical dev workflow is that after adding new properties/columns to your data models you would copy them into a new migration then run a tool to apply those migrations to the current database with the state of which migrations have been run is maintained in a `Migration` table.

> The more esoteric thought I've been having is about what I would called type-safe de/serialization. 

Not quite sure what this feature is exactly describing but the way you can control how parameterized values and mapped to the underlying RDBMS driver and how they're mapped back into your data models is using [Type Converters](https://litdb.dev/customize#type-converters). This should let you use custom types in your data models and then use Type Converters to convert them to/from your DB.

> Finally, I wanted to point out that $ is reserved in [Svelte](https://svelte.dev/).

Anyone can choose whatever alias they want, either when [exporting their connection](https://litdb.dev/#driver-usage) or when importing it. The examples use `$` since that's my personal preference (and what all our code bases use), but anyone can just use VS Codes/IDEs rename feature to rename it to anything else they prefer.

> Thanks for your awesome work!

Thanks for the kind words :) Hope you find litdb useful!

**Reply by ixxie** [reply]
Thanks for the thorough reply, I appreciate all the details you went into. A lot of the concepts are new to me and this provides me a lot of food for thought. I'm going to dig deeper into all of this when I revisit LitDB.

**Comment by alexanderniebuhr** [comment]
Can you add comparison with other solutions, e.g. Kysely?

**Reply by mythz** [reply]
I've never tried Kysely but from a quick glance at its homepage, its APIs look more stringly typed and APIs don't look natural, I prefer using SQL expressions with typed references.

The query builder also looks coupled to execution whereas litdb query builders and aren't coupled to a driver implementation, i.e. they're just used to generate SQL with parameters that could then be executed with any driver.

I prefer litdb's API and decoupling but I'm sure Kysely is also a great option.
